utils/comparable_bytes.cc

```diff
@@ -397,6 +406,127 @@ struct to_comparable_bytes_visitor {
+    //
+    // To compare, we need a normalized value, i.e. one with a sign, exponent and (0,1) mantissa.
+    // To avoid loss of precision, both exponent and mantissa need to be base-100.
+    // We cannot get this directly off the serialized bytes and the unscaled value has to be reconstructed first.
+    // After calculating such a mantissa and base 100 exponent, the value is encoded as follows :
+    //     - sign bit inverted + 0x40 + signed exponent length, where exponent is negated if value is negative
+    //     - zero or more exponent bytes (as given by length)
+    //     - 0x80 + first pair of decimal digits, negative if the decimal is negative and the digits are rounded to -inf
+    //     - zero or more 0x80 + pair of decimal digits, always positive
+    //     - trailing 0x00
+    // Zero is special-cased as 0x80.
+    //
+    // Because the trailing 00 cannot be produced from a pair of decimal digits (positive or not),
+    // no value can be a prefix of another.
+    //
+    // Encoding examples:
+    //    1.1    as       c1 = 0x80 (positive number) + 0x40 + (positive exponent) 0x01 (exp length 1)
+    //                    01 = exponent 1 (100^1)
+    //                    81 = 0x80 + 01 (0.01)
+    //                    8a = 0x80 + 10 (....10)   0.0110e2
+    //                    00
+    //    -1     as       3f = 0x00 (negative number) + 0x40 - (negative exponent) 0x01 (exp length 1)
+    //                    ff = exponent -1. negative number, thus 100^1
+    //                    7f = 0x80 - 01 (-0.01)    -0.01e2
+    //                    00
+    //    -99.9  as       3f = 0x00 (negative number) + 0x40 - (negative exponent) 0x01 (exp length 1)
+    //                    ff = exponent -1. negative number, thus 100^1
+    //                    1c = 0x80 - 100 (-1.00)
+    //                    8a = 0x80 + 10  (+....10) -0.999e2
+    //                    00
+    void operator()(const decimal_type_impl&) {
+        const int32_t scale = read_simple<int32_t>(serialized_bytes_view);
+        const boost::multiprecision::cpp_int unscaled_value =
+                value_cast<utils::multiprecision_int>(varint_type->deserialize_value(serialized_bytes_view));
+        if (unscaled_value.is_zero()) {
+            // special case for 0
+            out.write<uint8_t>(POSITIVE_DECIMAL_HEADER_MASK);
+            return;
+        }
+
+        // The big_decimal stores decimal of form unscaled_value * pow(10, -scale)
+        // A postive scale moves the decimal point to the left and a negative scale moves it to the right.
+        // The decimal 123.456 can be stored as unscaled value = 123456 and scale = 3.
+        // This needs to be normalized and converted to a mantissa that lies in [0.01, 1) and a base-100 exponent.
+        // such that the big_decimal value = mantissa [0.01, 1) * 100 ^ exponent.
+        // The decimal 123.456 becomes mantissa = 0.0123456 and base 100 exponent = 2 (i.e.) 0.0123456 * 100 ^ 2.
```

Comment:
This could be very expensive to compute, to the point where it can cause stalls.
That said I don't know what we can do about it, I don't see how else we can make decimals byte comparable and Cassandra already made a decision, we have to follow if we want to remain compatible.
Maybe we should discourage (or outright forbid) using decimals in keys altogether.

This doesn't have to block this PR.

Reply:
We can delay optimization of decimal until later. 

That said, it already caused huge trouble and can't be ignored. It will be easier to do it in a dedicated patch series.